web3/content_script.js

```diff
@@ -26,133 +25,142 @@ import {
   TRANSPORT_UNINITIALIZED,
   EV_WEB3_REQUEST, EV_WEB3_RESPONSE,
   EV_WEB3_ACCOUNTS_REQUEST, EV_WEB3_ACCOUNTS_RESPONSE,
-  EV_TOKEN, EV_SIGNER_BAR
+  EV_TOKEN, EV_SIGNER_BAR,
+  isEnabled
 } from '../shared';
 
-const script = document.createElement('script');
-script.src = chrome.extension.getURL('web3/inpage.js');
-document.documentElement.insertBefore(script, document.documentElement.childNodes[0]);
-
-const initPort = () => {
-  const port = chrome.runtime.connect({ name: 'web3' });
-  if (!port) {
-    return;
-  }
-
-  port.onMessage.addListener((msg) => {
-    const { id, err, payload } = msg;
-
-    // Inject iframe only if the page is using Web3
-    if (!err) {
-      injectIframe();
-    } else {
-      removeIframe(err);
+isEnabled()
+  .then((enabled) => {
+    if (enabled) {
+      main();
     }
-
-    window.postMessage({
-      type: EV_WEB3_RESPONSE,
-      id,
-      err,
-      payload
-    }, '*');
-  });
-
-  port.onDisconnect.addListener(() => {
-    port.isDisconnected = true;
   });
 
-  return port;
-};
-
-// process requests
-let port = initPort();
-window.addEventListener('message', (ev) => {
-  if (ev.source !== window) {
-    return;
-  }
-
-  if (!ev.data.type) {
-    return;
-  }
-
-  const { type } = ev.data;
+function main () {
+  const script = document.createElement('script');
+  script.src = chrome.extension.getURL('web3/inpage.js');
+  document.documentElement.insertBefore(script, document.documentElement.childNodes[0]);
```

Comment:
Hopefuly it will run fast enough despite being async. I will make a test page that only has a single `<script> console.log(web3)</script>` to check if it works correctly.

If it won't work we will need to inject `web3` every time, people can still disable the whole extension anyway (or we can mark `web3` as disconnected when we detect that it shouldn't be enabled)